### PR TITLE
cri_stats_provider: Implement removeTerminatedContainer correctly

### LINKED ISSUE #74310: initContainer metrics fail
**What happened**:
the metrics-server is unable to get metrics for pods with init containers.
See also: https://github.com/kubernetes-incubator/metrics-server/issues/213#issuecomment-465675386

**What you expected to happen**:
Metrics for pods with init containers show up.

**How to reproduce it (as minimally and precisely as possible)**:
create a pod with an init container. metrics are not gathered for that pod. 

**Environment**:
- Kubernetes version (use `kubectl version`):
1.13.2
- Runtime:
 containerd 1.2.2

### PR DESCRIPTION
**Pull request by littleroad**
The exited container in the summary (CPU and Memory are null) will cause
the metrics-server to skip the processing of the pod, which will make HPA
not work. Therefore, the exited container should be removed from the
summary.

However, the implementation of removeTerminatedContainer previously ignored
the difference between the runtimeapi.Container provided by cri and the
cadvisorapiv2.ContainerInfo provided by cadvisor, so it did not work as
expected.

From the cadvisor perspective, the cgroup directory for exited containers
no longer exists, so we don't have to filter them out. On the contrary, for
cri, since we have obtained the State of the containers, we only to filter
which is running.

Fixes: 3bd315dcb1cd ("Filter out duplicated container stats")
Signed-off-by: Lu Fengqi <lufq.fnst@cn.fujitsu.com>

**What type of PR is this?**

/kind bug

**What this PR does / why we need it**:
Reimplement removeTerminatedContainer correctly.

**Which issue(s) this PR fixes**:

Fixes #74310 

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

### CODE REVIEW

**Comment by WanLinghao** [comment]
@dashpole Seems a reasonable patch

**Comment by dashpole** [comment]
I think we have had issues in the past with inconsistent state in the container runtime, where there are more containers running than there should be, usually because of a bug with the container runtime.

Can you explain more about what problem this fixes?  Maybe with an example?

**Comment by kfox1111** [comment]
@dashpole see https://github.com/kubernetes-incubator/metrics-server/issues/213#issuecomment-465675386

**Comment by steven-sheehy** [comment]
I am not an expert, but I think @dashpole is correct. The existing logic was added in PR #54606 and is there to ensure duplicates are not counted. I think the existing logic should stay, but  containerMap construction in previous section should only include running containers.

/cc @Random-Liu @miaoyq

**Comment by littleroad** [comment]
@dashpole I apologize for missing the discussion about PR #54606. Since I only read the abbreviated commit message and function name, I didn't realize that this logic is a workaround for an existing runtime bug. 

This PR solves the problem that when the containerd is used, the metrics-server will skip the processing of the pod containing the init container and cause the HPA to fail.

Because even if the init container finishes its work and exits, cri will still report its stats to the kubelet. The kubelet will not filter it because there is no running container with the same name. Therefore, the following cpu and memory usage will make the metrics-server confused whether the stats of the container to be acquired successfully, so the stats statistics of the pod are discarded.
[commit e35b90310bea ("Make sure to skip fine-grained missing data")](https://github.com/kubernetes-incubator/metrics-server/commit/e35b90310beae0d7a3c30b1c9b4da5a1d3b1f2f7)

```
{
      "name": "istio-init",
      "startTime": "2019-02-06T15:17:38Z",
      "cpu": {
       "time": null
      },
      "memory": {
       "time": null
      }
[snip]
```

I will update this commit to keep the deduplication logic of the running container.

By the way, I am surprised that the runtime can't work as expected, so I will dive into the related discussion later.

**Comment by dashpole** [comment]
@littleroad thanks.  Ping this PR when you have made the changes.

File: pkg/kubelet/stats/cri_stats_provider.go [hunk 1]
```diff
@@ -567,34 +565,10 @@ func (p *criStatsProvider) cleanupOutdatedCaches() {
...
-	for _, refs := range containerMap {
-		if len(refs) == 1 {
-			result = append(result, refs[0])
-			continue
-		}
-		found := false
-		for i := 0; i < len(refs); i++ {
-			if refs[i].State == runtimeapi.ContainerState_CONTAINER_RUNNING {
-				found = true
-				result = append(result, refs[i])
```
Comment (littleroad) [inline pkg/kubelet/stats/cri_stats_provider.go]: After I revisited this code, I found that it did not actually perform any deduplication. I think it is indeed useless.
Suppose there are indeed multiple running containers with the same name. Since we didn't break the for loop when we set found to true, we will continue to append other running containers with the same name to result.

Comment (miaoyq) [reply pkg/kubelet/stats/cri_stats_provider.go]: This function does two things, as follows :
- Find all running containers created by k8s
- If the containers of one pod are all in `non-runing` state, find the last created container.

Comment (littleroad) [reply pkg/kubelet/stats/cri_stats_provider.go]: To be precise, the premise is that we no longer need containers that are not running. For the first thing, all we need to do is traverse the containers array to pick the running containers.

Comment (miaoyq) [reply pkg/kubelet/stats/cri_stats_provider.go]: See: https://github.com/kubernetes/kubernetes/issues/53514#issuecomment-334629943

```go
 // A ContainerInfo is considered to be of a terminated container if it has an 
 // older CreationTime and zero CPU instantaneous and memory RSS usage.
```
This only exclude duplicated terminated containers(**duplicated** and **terminated**), not **terminated**.

Comment (littleroad) [reply pkg/kubelet/stats/cri_stats_provider.go]: If you are willing to look at the problem that the PR #74336 is trying to solve, I think you will understand what I mean.

TL;DR: These terminated container stats will make the metrics-server not work as expected, now I want to modify this code to filter out the terminated containers.

Comment (miaoyq) [reply pkg/kubelet/stats/cri_stats_provider.go]: I know what you mean, but if filter out all terminated containers, that may cause other problems.  
/cc @Random-Liu @yguo0905

Comment (littleroad) [reply pkg/kubelet/stats/cri_stats_provider.go]: I do not think so. Taking cadvisor as an example, except for the uncleaned cgroup directory problem #47853, the containers are no longer tracked by cadvisor once they are terminated. After the terminated container is filtered out, the result returned by cri and cadvisor is equivalent. 

Please let me know if my understanding is not correct.

**Comment by littleroad** [comment]
> I think we have had issues in the past with inconsistent state in the container runtime, where there are more containers running than there should be, usually because of a bug with the container runtime.

@dashpole I am very interested in the runtime bug mentioned above. Could you provide me more information (such as issue number, or keyword)?

I only found #47853 that the cgroup directory was not cleaned up in time so cadvisor may report wrong container stats. This kind of problem is reasonable for cadvisor, but it seems strange for the runtime.

**Comment by dashpole** [comment]
reading back through https://github.com/kubernetes/kubernetes/pull/54606, https://github.com/kubernetes/kubernetes/issues/53514, and https://github.com/kubernetes/kubernetes/issues/47853, it looks like this was mainly an issue with cAdvisor, and not cri-based runtimes.  But based on https://github.com/kubernetes/kubernetes/issues/53514#issuecomment-334629943, this is more than just "fixing a bug".  It is actually intended that the CRI report stats for _all_ containers, including multiple runs of the same container.  The code here is to make sure we only include stats from the latest run of that container.

**Comment by littleroad** [comment]
> But based on [#53514 (comment)](https://github.com/kubernetes/kubernetes/issues/53514#issuecomment-334629943), this is more than just "fixing a bug". It is actually intended that the CRI report stats for _all_ containers, including multiple runs of the same container. The code here is to make sure we only include stats from the latest run of that container in the summary API.

Let's go back to the beginning.
1. The cri-containerd returns the stats of all existing containers. On the contrary, Cadvisor only returns running containers because of its implementation.
2. The kubelet doesn't filter out the terminated containers, only remove duplicated containers with the same name before providing the stat/summary API. 
3. The lack of CPU/memory usage stats for the terminated container causes metrics-server to skip this pod.

Based on the above, we have the following three solutions:
1. Make the cri-containerd skip the terminated containers.
From Issue #53514, @Random-Liu and @yujuhong think that the CRI should return the stats of all containers. 
2. Make the kubelet filter out the terminated containers.
IMHO, the summary API doesn't need to provide the stat of the terminated containers that without the successor. Since the Cadvisor doesn't provide all terminated containers stats at all, are there any summary API consumers who concern the terminated containers?
3. Make the metrics-server ignore the containers which lack CPU/memory usage.
The metrics-server doesn't get the container state from the summary api, so it can't simply filter out the terminated container. If there are other reasons why the container stats lost, I am afraid that metrics-server can't ignore the containers which lack usage statistics.

Any suggestions are welcome.

**Comment by Random-Liu** [comment]
Sorry, I just saw this.

There are several options for this:
* Option 1: Always report 0 cpu/memory usage for exited container;
* Option 2: Only report stats for running container; (Actually running container still has disk usage)
* Option 3: Treat nil value as 0 in metrics server.

I think metrics server should do option 3 anyway, it is a protection to the system.
And we can do option 1 on the kubelet side if people think it is still necessary with the metrics server fix.

**Comment by Random-Liu** [comment]
Mark 1.14 and critical urgent.

We may want to cherry-pick this fix as well.

**Comment by steven-sheehy** [comment]
@Random-Liu For option 3, the metrics-server is rejecting the stats [here](https://github.com/kubernetes-incubator/metrics-server/blob/92d8412788e27ee669d38f21f20bad5342211884/pkg/sources/summary/summary.go#L122). I'm not sure rejecting nil cpu/memory stats will work because of the following comment in the code:
```
// NB: we explicitly want to discard pods with partial results, since
// the horizontal pod autoscaler takes special action when a pod is missing
// metrics (and zero CPU or memory does not count as "missing metrics")
```
Overloading use of nil vs 0 doesn't sound like a feasible approach in metrics-server. Also, if you send 0 stats for terminate in kubelet, it would need to include a non-zero timestamp for metrics-server to accept it.

**Comment by dashpole** [comment]
Is the correct behavior then to exclude init containers from the summary API?  It seems the metrics server is essentially looking for all non-init containers to be present.  If one of the non-init containers is terminated, the metrics server will discard the metrics no matter how we report it.

**Comment by steven-sheehy** [comment]
Technically a long running initContainer could be running at the time metrics are requested, so not sure excluding all initContainers would be a good idea.
/cc @DirectXMan12

**Comment by Random-Liu** [comment]
> Is the correct behavior then to exclude init containers from the summary API? It seems the metrics server is essentially looking for all non-init containers to be present. If one of the non-init containers is terminated, the metrics server will discard the metrics no matter how we report it.

Regular `RestartNever` container can have 0 cpu/memory usage as well. If this is the case, let's enforce 0 cpu/memory usage to be set for CRI stats.

**Comment by littleroad** [comment]
Report 0 CPU/memory usage and set a valid timestamp seems enough to fix this problem. Please feel free to point out if there is any problem.

**Comment by Random-Liu** [comment]
Can we add a unit test case?

LGTM other than that.

**Comment by DirectXMan12** [comment]
the "missing metrics" logic referred to above is designed to counter cases where metrics are *actually* unknown (i.e. cases where a container has yet to have metrics recorded because it just started up and cases where metrics are missing because of a failed scrape).  For cases where we can be *certain* that a container is supposed to be using zero CPU (i.e. it was running, and is now stopped), returning zero should be fine.

> Technically a long running initContainer could be running at the time metrics are requested, so not sure excluding all initContainers would be a good idea since we'd still want the resource usage metrics to be accurate.

Agreed, you want stats for everything, unless you actively know that you don't know stats for that pod (if that makes any sort of sense).

**Comment by littleroad** [comment]
> Can we add a unit test case?

The unit test case has been added.

**Comment by Random-Liu** [comment]
@littleroad Can you send cherry-pick to 1.11, 1.12 and 1.13? Thanks!
